# DATA WRANGLING
## Data Gathering, Data Assessing, Data Cleaning

## 1.0 Data Gathering

We have ready files in our working directory in this project

In [1]:
# import libraries
import numpy as np
import pandas as pd

In [2]:
# load the data files on the notebook
adverse_reactions = pd.read_csv('adverse_reactions.csv')
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')

## 2.0 Data Assessing Data 

Inspect the data for two issues:
- Data Quality:- Issues to do with its content. Also called dirty data. They include issues like missing, duplicate, or incorrect.
- Tidiness:- Data that has tidiness issues has issues with its structure.

### 1. Data Quality
Some data quality metrics
1. Completeness - Is the data complete?
2. Validity - The data is complete but is it valid?
3. Accuracy - The data is complete and valid but is it accurate?
4. Consistency - The data is complete, valid & accurate. Is it in the same standard format.

### 2. Data tidiness
Some data tidiness requirements include
1. Each Variable forms a column.
2. Each Observation forms a row.
3. Each type of observational unit forms a table. 

#### Visual assessment.

In [3]:
# Display patients table
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

About patients columns.

`patient_id`: the unique identifier for each patient in the Master Patient Index (i.e. patient database) of the pharmaceutical company that is producing Auralin.

`assigned_sex`: the assigned sex of each patient at birth (male or female).

`given_name`: the given name (i.e. first name) of each patient.

`surname`: the surname (i.e. last name) of each patient.

`address`: the main address for each patient.

`city`: the corresponding city for the main address of each patient.

`state`: the corresponding state for the main address of each patient.

`zip_code`: the corresponding zip code for the main address of each patient.

`country`: the corresponding country for the main address of each patient (all United states for this clinical trial).

`contact`: phone number and email information for each patient.

`birthdate`: the date of birth of each patient (month/day/year). The inclusion criteria for this clinical trial is age >= 18 (there is no maximum age because diabetes is a growing problem among the elderly population).

`weight`: the weight of each patient in pounds (lbs).

`height`: the height of each patient in inches (in).

`bmi`: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m2 where kg is a person's weight in kilograms and m2 is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. The inclusion criteria for this clinical trial is 16 >= BMI >= 38.


In [4]:
# Display treatments table
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before. All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:

175 patients switched to Auralin for 24 weeks
175 patients continued using Novodra for 24 weeks
treatments columns:

`given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial.

`surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial.

`auralin`: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) and the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the international unit of measurement and the standard measurement for insulin.

`novodra`: same as above, except for patients that continued treatment with Novodra.

`hba1c_start`: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The HbA1c test measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.

`hba1c_end`: the patient's HbA1c level at the end of the last week of treatment.

`hba1c_change`: the change in the patient's HbA1c level from the start of treatment to the end, i.e., hba1c_start - hba1c_end. 

For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [5]:
# Display adverse reactions table.
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

adverse_reactions columns:

`given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra).

`surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra).

`adverse_reaction`: the adverse reaction reported by the patient.

#### Programmatic assessment

Patients table

In [6]:
# Search for the completeness of the data. patients info
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
#check for missing values in the address column 
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [8]:
patients.surname.value_counts()

Doe            6
Jakobsen       3
Taylor         3
Ogochukwu      2
Tucker         2
              ..
Casárez        1
Mata           1
Pospíšil       1
Rukavina       1
Onyekaozulu    1
Name: surname, Length: 466, dtype: int64

In [9]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [10]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [11]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [12]:
patients.tail()

patient_id assigned_sex given_name      surname  \
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
498       2530 Victoria Court   Milton Mills        ME    3852.0   
499  494 Clarksburg Park Road         Sedona        AZ   86341.0   
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [13]:
patients.sample()

patient_id assigned_sex given_name surname           address   city  \
158         159         male    Chibuzo   Okoli  4291 Patton Lane  Tulsa   

    state  zip_code        country                                   contact  \
158    OK   74136.0  United States  ChibuzoOkoli@einrot.com+1 (918) 971-5864   

     birthdate  weight  height   bmi  
158  6/14/1983   142.8      73  18.8

In [14]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

#### Programmatic assessment

Treatments table

In [15]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [16]:
treatments.auralin.isnull().sum()

0

In [17]:
sum(treatments.novodra.isnull())

0

In [18]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [19]:
treatments.tail()

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36

In [20]:
treatments.sample()

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
116     yasmin   silva  36u - 44u       -         8.88       8.46   

     hba1c_change  
116           NaN

In [21]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

#### Programmatic assessment

adverse reactions table

In [22]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [23]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [24]:
adverse_reactions.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top         berta  johnson     hypoglycemia
freq            1        2               19

In [25]:
# .value_counts on the adverse_reaction column of the adverse_reactions table.
adverse_reactions.adverse_reaction.value_counts

<bound method IndexOpsMixin.value_counts of 0     injection site discomfort
1                  hypoglycemia
2                  hypoglycemia
3                         cough
4             throat irritation
5                  hypoglycemia
6                  hypoglycemia
7                  hypoglycemia
8                  hypoglycemia
9     injection site discomfort
10                     headache
11                        cough
12                 hypoglycemia
13    injection site discomfort
14                 hypoglycemia
15                       nausea
16                 hypoglycemia
17                       nausea
18                 hypoglycemia
19                     headache
20                 hypoglycemia
21    injection site discomfort
22    injection site discomfort
23                 hypoglycemia
24    injection site discomfort
25                 hypoglycemia
26            throat irritation
27                 hypoglycemia
28                 hypoglycemia
29                 hypoglyce

In [26]:
# select the records in the patients table for patients that are from the city New York.
patients[patients['city'] == 'New York']

patient_id assigned_sex    given_name     surname               address  \
9            10       female        Sophie     Cabrera     3303 Anmoore Road   
35           36       female        Kamila    Pecinová  3558 Longview Avenue   
84           85       female         Nương          Vũ   465 Southern Street   
129         130       female       Rebecca    Jephcott      989 Wayback Lane   
142         143         male        Finley    Chandler  2754 Westwood Avenue   
152         153         male   Christopher    Woodward  3450 Southern Street   
188         189         male         Søren    Sørensen      2397 Bell Street   
213         214       female    Onyemaechi   Onwughara     685 Duncan Avenue   
215         216         male          John         Doe       123 Main Street   
229         230         male          John         Doe       123 Main Street   
237         238         male          John         Doe       123 Main Street   
244         245         male          John         Doe       123 Main Street   
247         248         male        Tuukka  Leppäluoto    1886 Bicetown Road   
251         252         male          John         Doe       123 Main Street   
263         264       female         Julia    Carvalho     3662 Shinn Street   
277         278         male          John         Doe       123 Main Street   
301         302       female  Onyekachukwu      Obinna    2970 Forest Avenue   
461         462         male        Cannan     Cabrera   2102 Geraldine Lane   

         city     state  zip_code        country  \
9    New York  New York   10011.0  United States   
35   New York  New York   10004.0  United States   
84   New York        NY   10001.0  United States   
129  New York        NY   10004.0  United States   
142  New York  New York   10001.0  United States   
152  New York        NY   10004.0  United States   
188  New York        NY   10011.0  United States   
213  New York        NY   10013.0  United States   
215  New York        NY   12345.0  United States   
229  New York        NY   12345.0  United States   
237  New York        NY   12345.0  United States   
244  New York        NY   12345.0  United States   
247  New York        NY   10011.0  United States   
251  New York        NY   12345.0  United States   
263  New York        NY   10036.0  United States   
277  New York        NY   12345.0  United States   
301  New York        NY   10004.0  United States   
461  New York        NY   10014.0  United States   

                                               contact   birthdate  weight  \
9        SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7   
35               718-501-0503KamilaPecinova@dayrep.com  12/23/1985   198.9   
84                  VuCamNuong@fleckens.hu516-720-5094    2/1/1981   138.2   
129            631-370-7406RebeccaJephcott@armyspy.com    8/1/1966   203.3   
142              516-740-5280FinleyChandler@dayrep.com  10/25/1936   150.9   
152  ChristopherWoodward@jourrapide.com+1 (516) 630...    9/4/1984   212.2   
188            SrenSrensen@superrito.com1 212 201 3108  12/31/1942   157.1   
213         917-622-9142OnyemaechiOnwughara@einrot.com    3/8/1989   131.1   
215                        johndoe@email.com1234567890    1/1/1975   180.0   
229                        johndoe@email.com1234567890    1/1/1975   180.0   
237                        johndoe@email.com1234567890    1/1/1975   180.0   
244                        johndoe@email.com1234567890    1/1/1975   180.0   
247           917-408-8855TuukkaLeppaluoto@teleworm.us    3/7/1978   211.0   
251                        johndoe@email.com1234567890    1/1/1975   180.0   
263  JuliaAzevedoCarvalho@superrito.com+1 (212) 782...   4/11/1931   171.8   
277                        johndoe@email.com1234567890    1/1/1975   180.0   
301         OnyekachukwuObinna@teleworm.us646-982-6609   1/24/1997   154.7   
461     646-289-4177CannanCabreraOrdonez@superrito.com  10/12/1980   209.7   

     height 

In [27]:
len(patients[patients['city']== 'New York'])

18

In [28]:
len(patients[patients['state']== 'New York'])

25

### Assessment summary
#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean yet)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames


#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## 3.0 Data Cleaning

Data cleaning tips

1. Make a copy of the dataframe first - This ensures you have an original copy within your notebook incase an important feature is lost while cleaning

2. Address completeness issues first. Address the missing data. This can be via imputing, dropping, or ignoring

3. Address all tidiness issues next. Ensure all data is tidy so that cleaning steps are not repeated.

4. Address all other data quality issues. Resolve issues to do with invalidity, inaccuracy and inconsistency.

#### 3.1 Make a copy of the data

In [29]:
# make copies of the data
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

#### 3.2  Address completeness issues. 
##### 3.2.1 Missing data

`treatments` Missing records (280 instead of 350)

##### Define
Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame.

##### Code

In [30]:
# load treatments_cut
treatments_cut = pd.read_csv('treatments_cut.csv')

# concat treatments cut and treatments_clean 
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index=True)

##### Test

In [31]:
# check info to acertain 350 rows
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


In [32]:
# check the tail
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

##### 3.2.2 Missing data 
##### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

##### Define
Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

##### Code

In [33]:
treatments_clean.hba1c_change = treatments_clean.hba1c_start-treatments_clean.hba1c_end

##### Test

In [34]:
# check for any null values in the hba1c_change column
treatments_clean.hba1c_change.isnull().sum()

0

In [35]:
# check the head
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

#### 3.3 Address Tidiness issues. 

##### 3.3.1 Contact column in `patients` table contains two variables: phone number and email

##### Define
Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

##### Code

In [36]:
# glimpse current state of contact column
patients_clean.contact.head()

0          951-719-9170ZoeWellish@superrito.com
1         PamelaSHill@cuvox.de+1 (217) 569-3204
2              402-363-6804JaeMDebord@gustr.com
3    PhanBaLiem@jourrapide.com+1 (732) 636-8246
4               334-515-7487TimNeudorf@cuvox.de
Name: contact, dtype: object

In [37]:
# Create phone_number and email columns by extracting data from contact column
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})',
                                                                     expand=True)

patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.][a-zA-Z]+)',
                                                             expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row. Drop the contacts column.
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [38]:
# confirm contact column no longer exists
patients_clean.columns


Index(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'birthdate', 'weight', 'height',
       'bmi', 'phone_number', 'email'],
      dtype='object')

In [39]:
# View samples of the phone_number column
patients_clean.phone_number.sample(20)

466    +1 (612) 589-1495
255         715-231-3508
126         508-739-5632
82          678-350-2390
38          978 460 9060
343         203-251-3573
448         216-502-3773
1      +1 (217) 569-3204
124    +1 (508) 526-3432
78          774 263 5317
460    +1 (510) 238-1151
30          303-910-2058
195         985-253-7891
221    +1 (267) 932-9852
385         434-361-8479
114         917-256-5329
232         858 587 9945
265    +1 (989) 826-7884
72          504-289-1386
329         813 434 8122
Name: phone_number, dtype: object

In [40]:
# View samples of the email column
patients_clean.email.sample(20)

389             YegorUspensky@fleckens.hu
415                MgbekeEkwueme@cuvox.de
322             LieuThanhMuoi@teleworm.us
317                NancyJParker@gustr.com
375             SofiaBHermansen@rhyta.com
401                  KasumiHori@gustr.com
330              VascovandeWiel@rhyta.com
95            SargentRuais@jourrapide.com
122                 TaHaBoi@superrito.com
94                  GinaACain@armyspy.com
334              EugeneMironov@dayrep.com
213        OnyemaechiOnwughara@einrot.com
314              RenzoLucchese@dayrep.com
340          MineaLindgren@jourrapide.com
272    KlementynaSokolowska@superrito.com
159            CoralieAllaire@armyspy.com
466             AndreaBrodahl@armyspy.com
476             BertaNapolitani@rhyta.com
5            RafaelCardosoCosta@gustr.com
295           AlbinoSchiavone@fleckens.hu
Name: email, dtype: object

In [41]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

##### 3.3.2 Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

##### Code

In [42]:
# glimpse the dataframe
treatments_clean.head(2)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   

   hba1c_change  
0          0.43  
1          0.47

In [43]:
# column that will remaina as they are id_vars=['given_name', 'surname','hba1c_start', 'hbaic_end', 'hba1c_change']
# put auralin & novadra in var_name = 'treatment_dose'
# put the values in value_name = 'dose'   
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment_dose', value_name='dose')

# pick rows with dose values
treatments_clean = treatments_clean[treatments_clean.dose != "-"]

# split the dose column into dose_start and dose_end
treatments_clean['dose_start'],treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str

# drop the dose column
treatments_clean = treatments_clean.drop('dose', axis=1)

C:\Users\Dave\AppData\Local\Temp\ipykernel_3488\3615292242.py:11: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_clean['dose_start'],treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str


##### Test

In [44]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment_dose  \
0   veronika    jindrová         7.63       7.20          0.43        auralin   
3       skye  gormanston         7.97       7.62          0.35        auralin   
6     sophia      haugen         7.65       7.27          0.38        auralin   
7      eddie      archer         7.89       7.55          0.34        auralin   
9       asia     woźniak         7.76       7.37          0.39        auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

##### 3.3.3 Adverse reaction should be part of the `treatments` table

##### Define
Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

##### Code

In [45]:
adverse_reactions_clean.head(2)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia

In [46]:
treatments_clean.head(2)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment_dose  \
0   veronika    jindrová         7.63       7.20          0.43        auralin   
3       skye  gormanston         7.97       7.62          0.35        auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u

In [47]:
# merge the two tables on given_name and surname
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name', 'surname'], how='left')

##### Test

In [48]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change  \
0       veronika    jindrová         7.63       7.20          0.43   
1           skye  gormanston         7.97       7.62          0.35   
2         sophia      haugen         7.65       7.27          0.38   
3          eddie      archer         7.89       7.55          0.34   
4           asia     woźniak         7.76       7.37          0.39   
..           ...         ...          ...        ...           ...   
345  christopher    woodward         7.51       7.06          0.45   
346        maret    sultygov         7.67       7.30          0.37   
347        lixue       hsueh         9.21       8.80          0.41   
348        jakob    jakobsen         7.96       7.51          0.45   
349        berta  napolitani         7.68       7.21          0.47   

    treatment_dose dose_start dose_end           adverse_reaction  
0          auralin        41u      48u                        NaN  
1          auralin        33u      36u                        NaN  
2          auralin        37u      42u                        NaN  
3          auralin        31u      38u                        NaN  
4          auralin        30u      36u                        NaN  
..             ...        ...      ...                        ...  
345        novodra        55u      51u                     nausea  
346        novodra        26u      23u                        NaN  
347        novodra        22u      23u  injection site discomfort  
348        novodra        28u      26u               hypoglycemia  
349        novodra        42u      44u  injection site discomfort  

[350 rows x 9 columns]

##### 3.3.4 given_name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the `patients` table, then convert these names to lower case to join with `treatments`. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

##### Code

In [49]:
patients_clean.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   

   height   bmi       phone_number                     email  
0      66  19.6       951-719-9170  ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204      PamelaSHill@cuvox.de

In [50]:
id_names = patients_clean.loc[:, ['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()

treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

##### Test

In [51]:
# check for patient_id. Note that given_name and surname should be dropped.
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment_dose dose_start dose_end  \
0           7.63       7.20          0.43        auralin        41u      48u   
1           7.97       7.62          0.35        auralin        33u      36u   
2           7.65       7.27          0.38        auralin        37u      42u   
3           7.89       7.55          0.34        auralin        31u      38u   
4           7.76       7.37          0.39        auralin        30u      36u   
..           ...        ...           ...            ...        ...      ...   
344         7.51       7.06          0.45        novodra        55u      51u   
345         7.67       7.30          0.37        novodra        26u      23u   
346         9.21       8.80          0.41        novodra        22u      23u   
347         7.96       7.51          0.45        novodra        28u      26u   
348         7.68       7.21          0.47        novodra        42u      44u   

              adverse_reaction  patient_id  
0                          NaN         225  
1                          NaN         242  
2                          NaN         345  
3                          NaN         276  
4                          NaN          15  
..                         ...         ...  
344                     nausea         153  
345                        NaN         420  
346  injection site discomfort         336  
347               hypoglycemia          25  
348  injection site discomfort         477  

[349 rows x 8 columns]

In [52]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

#### 3.4  Address Quality issues. 

##### 3.4.1 Data types
##### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert the zip code column's data type from a float to a string using `astype`, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

##### Code

In [53]:
# check data type
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
 13  phone_number  491 non-null    object 
 14  email         491 non-null    object 
dtypes: float64(3), int64(2), object(10)
memory usage: 59.1+ KB


In [54]:
patients_clean.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   

   height   bmi       phone_number                     email  
0      66  19.6       951-719-9170  ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204      PamelaSHill@cuvox.de

In [55]:
#convert to string data type
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [56]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

##### 3.4.2 Inaccuracy
##### Tim Neudorf height is 27 in instead of 72 in

##### Define
Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [57]:
patients_clean.height = patients_clean.height.replace(27, 72)

In [58]:
# Should be empty
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [59]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

##### 3.4.3 Inconsistency
#####  Full state names sometimes, abbreviations other times

##### Define
Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [60]:
# check for unique state names
patients_clean.state.unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

In [61]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

##### Test

In [62]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
MI    13
OK    13
LA    13
NJ    12
VA    11
WI    10
MS    10
AL     9
TN     9
IN     9
MN     9
NC     8
KY     8
WA     8
MO     7
NE     6
KS     6
ID     6
NV     6
SC     5
IA     5
CT     5
RI     4
ND     4
AR     4
AZ     4
ME     4
CO     4
MD     3
DE     3
SD     3
WV     3
OR     3
VT     2
MT     2
DC     2
AK     1
NM     1
NH     1
WY     1
Name: state, dtype: int64

In [63]:
patients_clean.state.unique()

array(['CA', 'IL', 'NE', 'NJ', 'AL', 'FL', 'NV', 'MO', 'NY', 'MI', 'TN',
       'VA', 'OK', 'GA', 'MT', 'MA', 'NM', 'LA', 'PA', 'CO', 'ME', 'WI',
       'SD', 'MN', 'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD',
       'AZ', 'TX', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH',
       'OR', nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

##### 3.4.4 Inaccuracy
##### Dsvid Gustafsson

##### Define
Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

##### Code

In [64]:
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [65]:
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

##### 3.4.5 Inconsistency in data types 
##### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

In [66]:
patients_clean.dtypes

patient_id        int64
assigned_sex     object
given_name       object
surname          object
address          object
city             object
state            object
zip_code         object
country          object
birthdate        object
weight          float64
height            int64
bmi             float64
phone_number     object
email            object
dtype: object

In [67]:
treatments_clean.head(2)

hba1c_start  hba1c_end  hba1c_change treatment_dose dose_start dose_end  \
0         7.63       7.20          0.43        auralin        41u      48u   
1         7.97       7.62          0.35        auralin        33u      36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242

##### Define
Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.


##### Code

In [68]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

##### Test

In [69]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [70]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment_dose    349 non-null    object 
 4   dose_start        349 non-null    int32  
 5   dose_end          349 non-null    int32  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int32(2), int64(1), object(2)
memory usage: 21.8+ KB


In [71]:
treatments_clean.head(2)

hba1c_start  hba1c_end  hba1c_change treatment_dose  dose_start  dose_end  \
0         7.63       7.20          0.43        auralin          41        48   
1         7.97       7.62          0.35        auralin          33        36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242

##### 3.4.6 Multiple phone number formats

##### Define
Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [72]:
patients_clean.phone_number.sample(20)

206         806-323-2210
15          276-225-1955
163         803-712-1180
368         617 830 7216
491         210-222-8684
112    +1 (757) 885-5669
471         503-417-1995
414         937-526-1152
252         978-243-8596
203         636-442-6946
4           334-515-7487
114         917-256-5329
394         903-939-1025
199         205 648 4197
302         601-296-7799
55          702 730 5584
158    +1 (918) 971-5864
66          760-287-8295
318         718 954 8136
335         765-359-0147
Name: phone_number, dtype: object

In [73]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '', regex=True).str.pad(11, fillchar='1')

##### Test

In [74]:
patients_clean.phone_number.sample(20)

446    16014180102
168    16508496900
92     16099148473
432    19792030438
19     15305328397
439    15153626574
26     13096718852
276    12082649063
326    15183790603
295    19168179960
18     14067752696
106    17655998847
15     12762251955
264            NaN
80     16466210640
298    13615335161
223    12088973897
95     13252824087
229    11234567890
37     16052046572
Name: phone_number, dtype: object

##### 3.4.7 Default John Doe data

##### Define
Remove the non-recoverable John Doe records from the `patients` table.

##### Code

In [75]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

In [76]:
# There should be no doe records
patients_clean.surname.value_counts()

Jakobsen       3
Taylor         3
Aranda         2
Tucker         2
Souza          2
              ..
Casárez        1
Mata           1
Pospíšil       1
Rukavina       1
Onyekaozulu    1
Name: surname, Length: 465, dtype: int64

In [77]:
# Should be no 123 Main Street records
patients_clean.address.value_counts()

2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
576 Brown Bear Drive        1
2272 Williams Avenue        1
                           ..
1066 Goosetown Drive        1
4291 Patton Lane            1
4643 Reeves Street          1
174 Lost Creek Road         1
3652 Boone Crockett Lane    1
Name: address, Length: 482, dtype: int64

##### 3.4.8 Multiple records for Jakobsen, Gersten, Taylor

##### Define
Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [78]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [79]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [80]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [81]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

##### 3.4.9 kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [82]:
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

##### Test

In [83]:
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values()

459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 494, dtype: float64